Connection State: Error Handling + Cleaning / refactoring /

BillCarsonFr opened the pull request:
Better reviewed commit per commit

BillCarsonFr committed:
refactoring: prep work extract to file + documentation

BillCarsonFr committed:
refactor: Remote / Publish Connection and constructor

BillCarsonFr committed:
Connection: add Connection state and handle error on start

BillCarsonFr committed:
Connection states tests

BillCarsonFr committed:
lint fixes

BillCarsonFr committed:
extract common test setup

BillCarsonFr committed:
ConnectionState: test livekit connection states

BillCarsonFr committed:
tests: end scope tests

BillCarsonFr committed:
fix lint errors

BillCarsonFr committed:
tests: Add publisher observable tests

BillCarsonFr committed:
test: Ensure scope for publishers observer

BillCarsonFr committed:
refactor: PublishConnection extract from giant constructor

BillCarsonFr committed:
tests: Publish connection states

BillCarsonFr committed:
Merge branch 'voip-team/rebased-multiSFU' into valere/multi-sfu/connection_states

BillCarsonFr committed:
Lint: fix all the lint errors

BillCarsonFr committed:
build: update lock file

BillCarsonFr committed:
lint: fix import order

BillCarsonFr committed:
prettier !

BillCarsonFr committed:
knip: remove dead code

BillCarsonFr committed:
fixup MediaView tests

BillCarsonFr committed:
test: Fix mediaView test, ,member is not optional anymore

Updated the test because now name will be the userId instead of default display name

BillCarsonFr committed:
esLint fix

BillCarsonFr committed:
Remove un-necessary transport field, already accessible from connection

BillCarsonFr committed:
test: Fix Audio render tests and added more

robintown commented on src/livekit/MatrixAudioRenderer.tsx:
```diff
@@ -32,8 +32,11 @@ export interface MatrixAudioRendererProps {
    * This list needs to be composed based on the matrixRTC members so that we do not play audio from users
    * that are not expected to be in the rtc session.
    */
+  // TODO: Why do we have this structure? looks like we only need the valid/active participants (not the room member or id)?
```
Just laziness, really. The CallViewModel needs the ID and room member internally to create the media tiles, so I took the same data and made it public for the purpose of audio rendering without making the public type any stricter. I've remedied this in: c96e81bfd375262be9c992cbd1885c7d878f5b8d

robintown commented on src/main.tsx:
```diff
@@ -11,7 +11,6 @@ Please see LICENSE in the repository root for full details.
 // dependency references.
 import "matrix-js-sdk/lib/browser-index";
 
-import { StrictMode } from "react";
```
Oops, I didn't mean to commit my temporary disabling of strict mode. e346c8c148bd6aad302598d57f16f02dc91cf468

robintown committed:
Re-enable React strict mode

robintown committed:
Simplify type of audio participants exposed from CallViewModel

robintown committed:
Remove dead MuteStates file

It's been replaced by a refactored RxJS version living in src/state.

robintown commented on src/state/MuteStates.ts:
```diff
@@ -132,6 +138,7 @@ class MuteState<Label, Selected> {
   ) {}
 }
 
+// TODO there is another MuteStates in src/room/MuteStates.tsx ?? why
```
Dead code, we can remove it. 5da780ed301bc6e259ba9fb52c7b92b3aef88525

robintown commented on src/state/Async.ts:
```diff
@@ -5,16 +5,10 @@ SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
+import { catchError, from, map, type Observable, of, startWith } from "rxjs";
 
+// TODO where are all the comments? ::cry::
+// There used to be an unitialized state!, a state might not start in loading
```
IMO the "uninitialized" state doesn't belong here. In cases where it might be tempting to use "uninitialized", I'd like it if we explicitly add such a case into our types as needed (`Async<T>` → `Async<T | undefined>`) or keep an an eye out for better ways to model the problem. With just these three states, there's a very clear mapping between async values and Promises.

I added hopefully a decent amount of comments: b1d143720aa322d9f263c4e3736a0dd5ebc0d468

robintown committed:
Add comments to Async

robintown commented on src/state/Connection.ts:
```diff
@@ -10,293 +10,246 @@ import {
-    >,
-    e2eeLivekitOptions: E2EEOptions | undefined,
-    livekitRoom: LivekitRoom | undefined = undefined,
+  /**
+   * The focus server to connect to.
+   */
+  public readonly localTransport: LivekitTransport;
```
Why is this called `localTransport` and not `transport`? It's often remote.

robintown replied:
ea17ed7253cb47adb29475732bace0486aff0cb7

robintown committed:
Correct / document some missing bits in tests

robintown committed:
Fix doc comment typo

robintown committed:
Restore a hidden 'null' state for the local transport/connection

robintown commented on src/state/CallViewModel.ts:
```diff
@@ -558,48 +563,51 @@ export class CallViewModel extends ViewModel {
   /**
    * The transport over which we should be actively publishing our media.
    */
-  private readonly localTransport$: Behavior<Async<LivekitTransport> | null> =
+  private readonly localTransport$: Behavior<Async<LivekitTransport>> =
     this.scope.behavior(
       this.transports$.pipe(
         map((transports) => transports?.local ?? null),
         distinctUntilChanged(deepCompare),
       ),
     );
```
For the record this can be null, it's just that `deepCompare` defaults its type parameter to `any` which ended up hiding the type error here. Restored in: 3691e7120d572082af5560aa5539aa02a5eb6906

robintown committed:
Remove unused useIsEarpiece hook

robintown committed:
Remove MockedObject from mockMediaDevices type signature

robintown commented on src/utils/test.ts:
```diff
@@ -416,13 +418,15 @@ export const deviceStub = {
     videoInput: deviceStub,
     ...data,
-  } as MediaDevices;
+  } as MediaDevices);
```
This object isn't a `Vitest.MockedObject` in the sense that functions on it don't have the `.mockClear()`, `.mockReturnValue()` etc. methods available, so we shouldn't assert that it is. dcc3ab641f59ae986edca267d9b7787ed9c546fa

robintown committed:
Remove local participant case (now enforced by types) from audio tests

robintown commented on src/livekit/MatrixAudioRenderer.test.tsx:
```diff
@@ -48,42 +57,148 @@ vi.mock("@livekit/components-react", async (importOriginal) => {
+      "@alice:DEV1",
+    ],
+    expectedAudioTracks: 2,
+  },
```
For the record I'm removing this test case because I managed to make the prop types of MatrixAudioRenderer strict enough so that it only accepts RemoteParticipants. 00daf834b655140d402515beb52505946e79b7d0

robintown commented on src/state/Connection.ts:
```diff
@@ -10,293 +10,246 @@ import {
+
+export type FocusConnectionState =
+  | { state: "Initialized" }
+  | { state: "FetchingConfig"; focus: LivekitTransport }
+  | { state: "ConnectingToLkRoom"; focus: LivekitTransport }
+  | { state: "PublishingTracks"; focus: LivekitTransport }
+  | { state: "FailedToStart"; error: Error; focus: LivekitTransport }
```
I made an attempt in my code to stick to the new term 'transport' and phase out usage of 'focus'. Do you think it makes sense to continue that here?

robintown replied:
2dc61346061b5078c46a345784b62f456bd43bbb

robintown committed:
Fix focus connection state typo, simplify its initialization

robintown commented on src/state/Connection.ts:
```diff
@@ -10,293 +10,246 @@ import {
+        state: "FetchingConfig",
+        focus: this.localTransport,
+      });
+      // TODO could this be loaded earlier to save time?
+      const { url, jwt } = await this.getSFUConfigWithOpenID();
```
Oh yeah, good idea. I think it's only safe for our users' privacy to do so for the *local* connection though. (We shouldn't leak to untrusted remote servers whether you've left your client sitting on the lobby screen of a video room, or something. Only once there is a clear intent to join the call should we fetch the JWT.)

robintown replied:
9f4e99310bd6ae5f0e4ff86450f2ab4757103cf0

robintown committed:
Update outdated comment

robintown commented on src/state/Connection.ts:
```diff
@@ -10,293 +10,246 @@ import {
             )
             // Pair with their associated LiveKit participant (if any)
-            .map((membership) => {
+            // Uses flatMap to filter out memberships with no associated rtc participant ([])
+            .flatMap((membership) => {
               const id = `${membership.sender}:${membership.deviceId}`;
               const participant = participants.find((p) => p.identity === id);
-              return { participant, membership };
+              return participant ? [{ participant, membership }] : [];
             }),
```
I think something went wrong here with the merge. If you take a look at 1820cac3f66e9b9801209990d5ad5469ad21e97a I was trying to make sure that even RTC memberships with no associated LiveKit participant are included in `publishingParticipants$`. (One could add them back in CallViewModel, but it seemed way easier and more efficient to just avoid filtering them out here.)

The type that I was aiming for was:

```typescript
Connection.publishingParticipants$: Behavior<{
  participant: RemoteParticipant | undefined;
  membership: CallMembership;
}[]>
```

Once we get the CallViewModel tests back up and running we'll be back to a state where this can't regress without a red flag.

robintown replied:
a6c4fb41483c005f839c2c91b726b62dea68fe85

robintown commented on src/state/Connection.ts:
```diff
@@ -10,293 +10,246 @@ import {
-        });
-        for (const track of tracks) {
-          await this.livekitRoom.localParticipant.publishTrack(track);
+    scope
+      .behavior<ConnectionState>(connectionStateObserver(this.livekitRoom))
+      .subscribe((connectionState) => {
+        const current = this._focusedConnectionState$.value;
+        // Only update the state if we are already connected to the LiveKit room.
+        if (current.state === "ConnectedToLkRoom") {
+          this._focusedConnectionState$.next({
+            state: "ConnectedToLkRoom",
+            connectionState,
+            focus: current.focus,
+          });
```
Perhaps a more elegant way to do this would be to change the type to:

```diff
 {
   state: "ConnectedToLkRoom";
-  connectionState: ConnectionState;
+  connectionState: Observable<ConnectionState>;
   focus: LivekitTransport;
 }
```

Then (on line 111) you can just write:

```typescript
this._focusedConnectionState$.next({
  state: "ConnectedToLkRoom",
  focus: this.localTransport,
  connectionState: connectionStateObserver(this.livekitRoom),
});
```

and remove this line 212 code.

robintown replied:
b0eb566a4f959503c8be8d4c2a03c227553041fd

robintown commented on src/state/PublishConnection.ts:
```diff
@@ -0,0 +1,277 @@
+
+    super(room, args);
+
+    // Setup track processor syncing (blur)
+    this.observeTrackProcessors(scope, room, trackerProcessorState$);
+    // Observe mute state changes and update LiveKit microphone/camera states accordingly
+    this.observeMuteStates(scope);
+    // Observe media device changes and update LiveKit active devices accordingly
+    this.observeMediaDevices(scope, devices, controlledAudioDevices);
+
+    this.workaroundRestartAudioInputTrackChrome(devices, scope);
```
Nice, this cleaner constructor is also a welcome change

robintown commented on src/state/PublishConnection.ts:
```diff
@@ -0,0 +1,277 @@
+      }
+      return this.livekitRoom.localParticipant.isCameraEnabled;
+    });
+    scope.onEnd(() => {
+      this.muteStates.audio.unsetHandler();
+      this.muteStates.video.unsetHandler();
+    });
```
In 86fb026be86d7ea0fa904babd920b6e1d0b9f1c0 I moved these calls to `unsetHandler` to the `stop` method as `scope.onEnd` happens much later than what we need for SFU-hopping. I believe they should stay there in `stop`.

robintown replied:
Actually, scratch that, we should just give the Connection class a proper ObservableScope of its own that ends at the right time (when CallViewModel is about to drop the Connection object).

robintown replied:
I can work on that.

robintown commented on src/state/PublishConnection.ts:
```diff
@@ -0,0 +1,277 @@
+
+    if (this.stopped) return;
+
+    // TODO this can throw errors? It will also prompt for permissions if not already granted
+    const tracks = await this.livekitRoom.localParticipant.createTracks({
+      audio: this.muteStates.audio.enabled$.value,
+      video: this.muteStates.video.enabled$.value,
+    });
+    if (this.stopped) return;
+    for (const track of tracks) {
+      // TODO: handle errors? Needs the signaling connection to be up, but it has some retries internally
+      // with a timeout.
+      await this.livekitRoom.localParticipant.publishTrack(track);
+      if (this.stopped) return;
+      // TODO: check if the connection is still active? and break the loop if not?
+    }
```
We should restore this diff from 86fb026be86d7ea0fa904babd920b6e1d0b9f1c0:

```diff
@@ -130,23 +131,35 @@ export class PublishConnection extends Connection {
     if (!this.stopped) await this.livekitRoom.connect(url, jwt);
 
     if (!this.stopped) {
-      const tracks = await this.livekitRoom.localParticipant.createTracks({
-        audio: this.muteStates.audio.enabled$.value,
-        video: this.muteStates.video.enabled$.value,
-      });
-      for (const track of tracks) {
-        await this.livekitRoom.localParticipant.publishTrack(track);
+      // TODO-MULTI-SFU: Prepublish a microphone track
+      const audio = this.muteStates.audio.enabled$.value;
+      const video = this.muteStates.video.enabled$.value;
+      // createTracks throws if called with audio=false and video=false
+      if (audio || video) {
+        const tracks = await this.livekitRoom.localParticipant.createTracks({
+          audio,
+          video,
+        });
+        for (const track of tracks) {
+          await this.livekitRoom.localParticipant.publishTrack(track);
+        }
       }
     }
   }
```

(by which I mean just the `TODO` and the `audio || video` guard.)

robintown replied:
95069fd3fafd3cd40fb2b48edb061da50796e8db

robintown committed:
Clean up subscriptions in Connection tests

robintown reviewed (commented):
Very nice work with fixing things up as you go. For the sake of getting our two branches merged into one as soon as possible, I went ahead and applied some changes directly for things which seemed hopefully uncontrovertial - if you like you can have a look through them. The rest of my comments are mostly about things from my branch that seem to have been lost in the previous merge conflicts.

I did also confirm that the app runs and connects to an SFU successfully.

robintown committed:
Remove outdated comment

BillCarsonFr committed:
fix: use correct TestEachFunction

BillCarsonFr committed:
Simplify AudioRenderer and add more tests

BillCarsonFr committed:
test: Fix mute test, behavior change from setMuted to setAudioEnabled

useCallViewKeyboardShortcuts() changed a param from `setMicrophoneMuted` to `setAudioEnabled`, the boolean arg of the callback is inverse tht it used to be

BillCarsonFr committed:
Test: Fix mocking to fix failing tests

BillCarsonFr committed:
test: MISSING_MATRIX_RTC_FOCUS renamed as MISSING_MATRIX_RTC_TRANSPORT

toger5 committed:
Merge branch 'livekit' into voip-team/rebased-multiSFU

Signed-off-by: Timo K <toger5@hotmail.de>

BillCarsonFr committed:
Merge branch 'voip-team/rebased-multiSFU' into valere/multi-sfu/connection_states

BillCarsonFr committed:
post merge lint fixes

BillCarsonFr committed:
test: fix additional test with proper mutestate fix

BillCarsonFr committed:
Merge branch 'voip-team/rebased-multiSFU' into valere/multi-sfu/connection_states

BillCarsonFr committed:
prettier fix

BillCarsonFr committed:
post merge fixes (js-sdk changes)
